packages/@aws-cdk/aws-codebuild/lib/build-spec.ts

```diff
@@ -96,41 +96,62 @@ class YamlBuildSpec extends BuildSpec {
 }
 
 /**
- * Merge two buildspecs into a new BuildSpec
+ * Merge two buildspecs into a new BuildSpec by doing a deep merge
  *
- * NOTE: will currently only merge commands, not artifact
- * declarations, environment variables, secrets, or any
- * other configuration elements.
+ * We decided to disallow merging of artifact specs, which is
+ * actually impossible since we can't merge two buildspecs with a
+ * single primary output into a buildspec with multiple outputs.
+ * In case of multiple outputs they must have identifiers but we won't have that information.
  *
- * Internal for now because it's not complete/good enough
- * to expose on the objects directly, but we need to it to
- * keep feature-parity for Project.
- *
- * @internal
- */
+ * In case of test reports we replace the whole object with the RHS (instead of recursively merging)
+*/
 export function mergeBuildSpecs(lhs: BuildSpec, rhs: BuildSpec): BuildSpec {
   if (!(lhs instanceof ObjectBuildSpec) || !(rhs instanceof ObjectBuildSpec)) {
     throw new Error('Can only merge buildspecs created using BuildSpec.fromObject()');
   }
 
-  return new ObjectBuildSpec(copyCommands(lhs.spec, rhs.spec));
-}
-
-/**
- * Extend buildSpec phases with the contents of another one
- */
-function copyCommands(buildSpec: any, extend: any): any {
-  if (buildSpec.version === '0.1') {
+  if (lhs.spec.version === '0.1') {
     throw new Error('Cannot extend buildspec at version "0.1". Set the version to "0.2" or higher instead.');
   }
+  if (lhs.spec.artifacts && rhs.spec.artifacts) {
+    // We decided to disallow merging of artifact specs, which is
+    // actually impossible since we can't merge two buildspecs with a
+    // single primary output into a buildspec with multiple outputs.
+    // In case of multiple outputs they must have identifiers but we won't have that information.
+    throw new Error('Only one build spec is allowed to specify artifacts.');
+  }
+
+  const merged = mergeDeep(lhs.spec, rhs.spec);
+
+  // In case of test reports we replace the whole object with the RHS (instead of recursively merging)
+  if (lhs.spec.reports && rhs.spec.reports) {
+    merged.reports = { ...lhs.spec.reports, ...rhs.spec.reports };
+  }
+
+  return new ObjectBuildSpec(merged);
+}
+
+function mergeDeep(lhs: any, rhs: any): any {
+  lhs = JSON.parse(JSON.stringify(lhs));
```

Comment:
I'd prefer doing a fully recursive function here that type-analyzes, of the form:

```ts
function mergeDeep(lhs, rhs) {

  if (Array.isArray(lhs) && Array.isArray(rhs)) {
    return [...lhs, ...rhs];
  }
  if (Array.isArray(lhs) || Array.isArray(rhs)) { return rhs; }

  if (isObject(lhs) && isObject(rhs)) {
    const ret: any = {...lhs};
    for (const k of Object.keys(rhs)) {
      ret[k] = k in lhs ? mergeDeep(lhs[k], rhs[k]) : rhs[k];
    }
    return ret;
  }

  return rhs;
}

```